assets/js/SignPanel.tsx

```diff
@@ -174,25 +151,28 @@ function SignPanel({
   currentTime,
   line,
   setConfigs,
-  signConfig,
+  signConfig: initialSignConfig,
   realtimeId,
   signGroup,
   ungroupSign,
   readOnly,
 }: SignPanelProps): JSX.Element {
-  const [staticLine1, setStaticLine1] = React.useState(signConfig.line1 || '');
-  const [staticLine2, setStaticLine2] = React.useState(signConfig.line2 || '');
+  const [signConfig, setSignConfig] = React.useState(initialSignConfig);
   const [customChanges, setCustomChanges] = React.useState(false);
   const [initialTime] = React.useState(currentTime);
   const [confirmingUngroup, setConfirmingUngroup] = React.useState(false);
 
+  React.useEffect(() => {
+    setSignConfig(initialSignConfig);
+  }, [setSignConfig, initialSignConfig]);
```

Comment:
I'm not sure of the reasoning here but it looks like this will set the sign config back to the "initial" (passed-in) config whenever that initial config changes. If you're in the middle of changing a sign's config, and it changes on its own due to a server update, wouldn't this wipe out your changes?

Reply:
The thing that prompted it was needing the sign panel to update when the "All to Auto/Off/Headway" was pressed. Since the function won't re-render when the props change automatically - only the state - you would click "All to Auto" and the signs wouldn't respond. So basically, yeah, by design it blows away the state when the server and props change, and it would wipe out changes if someone was doing it at the same time, but I couldn't think of a simple way to handle that scenario, and it seemed pretty unlikely to me.

Reply:
If we're sure this will only happen when someone else is editing _the same sign_ at the same time (and not just editing any sign), I'd agree it would be uncommon enough.

Reply:
Ack, good catch! `initialSignConfig` changes at a reference level if any sign is edited. I pushed up a commit that wraps the `setSignConfig(...)` call in `if (!hasCustomChanges)`, so if the form is open / there's unsaved changes, they won't get blown away.

Reply:
That's probably a decent compromise. Though, if I push "All to X", does that reset any custom changes I made in my own browser? Because in that case I might expect it to... but I guess there are edge cases everywhere in the current "concurrency" model and we can't really catch them all without a larger refactoring.

Reply:
Heh, no it leaves the form alone. But playing around with it, I think that feels fine. Say the sign is Off, and you open up the "Custom form", and someone clicks "All to Auto", that sign's config will change, and so the user will start to see the realtime-predictions on the preview of the sign. Meanwhile, their form will still be open, with their changes present and the "Apply" button enabled with the little "there are changes asterisk". In other words, it will be in exactly the state it would be in if someone had done "All to Auto" *first* and *then* someone went in and started changing the custom text. At that point, they can either cancel, and the sign will go back to Auto, or click Apply, and their changes will take effect. I think it works out pretty well.